File: scripts/backfill-history.py
#!/usr/bin/env python3
"""
Promptbook — Historic Data Sync (Backfill)

Scans local Claude Code JSONL transcripts and submits aggregate stats
as backfill Build Cards. Never sends prompts, code, or file paths —
only aggregate stats.

Usage:
  python3 backfill-history.py [--days N] [--dry-run] [--project DIR]

Options:
  --days N       How many days of history to backfill (default: 30)
  --dry-run      Parse and print stats without submitting to the API
  --project DIR  Only backfill sessions from this project directory
"""

import json
import sys
import time
import argparse
from pathlib import Path
from datetime import datetime
from collections import Counter
from typing import Optional


def find_jsonl_files(days: int, project_filter: Optional[str] = None, before: Optional[str] = None) -> list:
    """Find JSONL session files from the last N days, optionally before a cutoff date."""
    claude_dir = Path.home() / ".claude" / "projects"
    if not claude_dir.exists():
        print(f"No Claude Code projects found at {claude_dir}", file=sys.stderr)
        return []

    after_cutoff = time.time() - (days * 86400)
    before_cutoff = None
    if before:
        before_cutoff = datetime.fromisoformat(before.replace("Z", "+00:00")).timestamp()

    files = []

    for project_dir in claude_dir.iterdir():
        if not project_dir.is_dir():
            continue

        if project_filter:
            dir_name = project_dir.name
            decoded_project = dir_name.rsplit("-", 1)[-1] if "-" in dir_name else dir_name
            if project_filter.lower() not in dir_name.lower() and project_filter.lower() != decoded_project.lower():
                continue

        for jsonl_file in project_dir.glob("*.jsonl"):
            mtime = jsonl_file.stat().st_mtime
            if mtime >= after_cutoff and (before_cutoff is None or mtime < before_cutoff):
                files.append(jsonl_file)

    return sorted(files, key=lambda f: f.stat().st_mtime)


def derive_project_name(project_dir_name: str) -> str:
    """Derive a human-readable project name from the encoded directory name."""
    # Format: -Users-foo-Desktop-myproject → myproject
    parts = project_dir_name.split("-")
    # Filter out empty strings and find the last meaningful part
    meaningful = [p for p in parts if p]
    return meaningful[-1] if meaningful else project_dir_name


def parse_session(jsonl_path: Path) -> Optional[dict]:
    """Parse a JSONL transcript and extract aggregate stats."""
    session_id = jsonl_path.stem
    project_dir_name = jsonl_path.parent.name
    project_name = derive_project_name(project_dir_name)

    entries = []
    try:
        with open(jsonl_path, "r") as f:
            for line in f:
                line = line.strip()
                if not line:
                    continue
                try:
                    entries.append(json.loads(line))
                except json.JSONDecodeError:
                    continue
    except (IOError, OSError) as e:
        print(f"  Skipping {jsonl_path.name}: {e}", file=sys.stderr)
        return None

    if not entries:
        return None

    # Extract timestamps
    timestamps = []
    for entry in entries:
        ts = entry.get("timestamp")
        if ts and isinstance(ts, str):
            try:
                timestamps.append(datetime.fromisoformat(ts.replace("Z", "+00:00")))
            except ValueError:
                continue

    if len(timestamps) < 2:
        return None

    start_time = min(timestamps)
    end_time = max(timestamps)
    build_time_seconds = int((end_time - start_time).total_seconds())

    if build_time_seconds < 5:
        return None  # Too short to be a real session

    # Count prompts: only user entries with actual text content (real human prompts).
    # Empty type=user entries are tool results, hook responses, etc. — not prompts.
    def is_real_prompt(entry):
        if entry.get("type") != "user":
            return False
        msg = entry.get("message", {})
        content = msg.get("content") if isinstance(msg, dict) else None
        if isinstance(content, str):
            return len(content.strip()) > 0
        if isinstance(content, list):
            return any(
                isinstance(b, dict) and b.get("type") == "text"
                and isinstance(b.get("text"), str) and len(b["text"].strip()) > 0
                for b in content
            )
        return False

    prompt_count = sum(1 for e in entries if is_real_prompt(e))

    # Extract token usage from assistant messages
    total_input = 0
    total_output = 0
    total_cache_creation = 0
    total_cache_read = 0
    model = None
    tool_usage: Counter = Counter()
    file_extensions: Counter = Counter()

    for entry in entries:
        if entry.get("type") != "assistant":
            continue

        msg = entry.get("message", {})

        # Get model from first assistant message
        if model is None and msg.get("model"):
            model = msg["model"]

        # Sum token usage
        usage = msg.get("usage", {})
        total_input += usage.get("input_tokens", 0)
        total_output += usage.get("output_tokens", 0)
        total_cache_creation += usage.get("cache_creation_input_tokens", 0)
        total_cache_read += usage.get("cache_read_input_tokens", 0)

        # Count tool usage
        for block in msg.get("content", []):
            if block.get("type") == "tool_use":
                tool_name = block.get("name", "Unknown")
                tool_usage[tool_name] += 1

                # Extract file extensions from tool inputs (Read, Edit, Write)
                tool_input = block.get("input", {})
                file_path = tool_input.get("file_path") or tool_input.get("path") or ""
                if file_path and "." in file_path:
                    ext = "." + file_path.rsplit(".", 1)[-1].lower()
                    if len(ext) <= 10:  # Reasonable extension length
                        file_extensions[ext] += 1

    if model is None:
        return None  # No assistant messages — not a real session

    # Calculate total tokens (cache reads excluded per cost-estimate.ts logic)
    total_tokens = total_input + total_cache_creation + total_output

    if total_tokens == 0 and prompt_count == 0:
        return None  # Empty session

    # Determine primary language from file extensions
    language = "Unknown"
    ext_to_lang = {
        ".ts": "TypeScript", ".tsx": "TypeScript",
        ".js": "JavaScript", ".jsx": "JavaScript",
        ".py": "Python",
        ".rs": "Rust",
        ".go": "Go",
        ".rb": "Ruby",
        ".java": "Java",
        ".cpp": "C++", ".c": "C", ".h": "C",
        ".cs": "C#",
        ".swift": "Swift",
        ".kt": "Kotlin",
        ".php": "PHP",
        ".sh": "Shell", ".bash": "Shell", ".zsh": "Shell",
        ".sql": "SQL",
        ".html": "HTML", ".css": "CSS", ".scss": "CSS",
        ".md": "Markdown",
        ".json": "JSON", ".yaml": "YAML", ".yml": "YAML",
        ".toml": "TOML",
    }
    if file_extensions:
        # Find the most common non-config extension
        for ext, _ in file_extensions.most_common():
            if ext in ext_to_lang and ext not in (".json", ".yaml", ".yml", ".toml", ".md"):
                language = ext_to_lang[ext]
                break
        else:
            # Fall back to most common extension even if config
            top_ext = file_extensions.most_common(1)[0][0]
            language = ext_to_lang.get(top_ext, "Unknown")

    return {
        "session_id": session_id,
        "project_name": project_name,
        "model": model,
        "ai_tool": "claude-code",
        "start_time": start_time.isoformat(),
        "end_time": end_time.isoformat(),
        "build_time_seconds": build_time_seconds,
        "prompt_count": prompt_count,
        "lines_changed": None,  # Not available from JSONL — omit, don't show 0
        "language": language,
        "status": "completed",
        "total_tokens": total_tokens,
        "input_tokens": total_input,
        "output_tokens": total_output,
        "cache_creation_input_tokens": total_cache_creation,
        "cache_read_input_tokens": total_cache_read,
        "source": "backfill",
        "source_metadata": {
            "file_extensions": dict(file_extensions.most_common(50)),
            "tool_usage_summary": dict(tool_usage.most_common(50)),
        },
    }


def upload_batch(sessions: list, api_url: str, api_key: str, days: int) -> Optional[str]:
    """Upload sessions to the backfill batch endpoint. Returns batch_id."""
    import urllib.request
    import urllib.error

    # Upload in chunks of 50
    chunk_size = 50
    batch_id = None

    for i in range(0, len(sessions), chunk_size):
        chunk = sessions[i:i + chunk_size]
        payload = json.dumps({
            "days_scanned": days,
            "sessions": chunk,
        }).encode("utf-8")

        req = urllib.request.Request(
            f"{api_url}/api/backfill/upload",
            data=payload,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            },
            method="POST",
        )

        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                body = json.loads(resp.read().decode("utf-8"))
                if batch_id is None:
                    batch_id = body.get("batch_id")
                uploaded = body.get("uploaded", 0)
                dupes = body.get("duplicates", 0)
                print(f"  Uploaded {uploaded} sessions ({dupes} duplicates)", file=sys.stderr)
        except urllib.error.HTTPError as e:
            err_body = e.read().decode("utf-8", errors="replace")
            print(f"  Upload failed (HTTP {e.code}): {err_body}", file=sys.stderr)
            # Stop on auth or server errors — retrying won't help
            if e.code in (401, 403, 500):
                print("  Stopping due to server error.", file=sys.stderr)
                break
        except Exception as e:
            print(f"  Upload failed: {e}", file=sys.stderr)
            break

        # Brief pause between chunks for rate limiting
        if i + chunk_size < len(sessions):
            time.sleep(2)

    return batch_id


def main():
    parser = argparse.ArgumentParser(description="Backfill Claude Code session history to Promptbook")
    parser.add_argument("--days", type=int, default=30, help="Days of history to backfill (default: 30)")
    parser.add_argument("--dry-run", action="store_true", help="Parse and print without submitting")
    parser.add_argument("--json", action="store_true", help="Output parsed sessions as JSON to stdout")
    parser.add_argument("--api-url", type=str, help="API URL (overrides config)")
    parser.add_argument("--api-key", type=str, help="API key (overrides config)")
    parser.add_argument("--project", type=str, help="Only backfill sessions from this project")
    parser.add_argument("--before", type=str, help="Only include sessions before this ISO date (e.g. 2026-03-10)")
    args = parser.parse_args()

    needs_config = not args.dry_run and not args.json

    # Load config
    config_path = Path.home() / ".promptbook" / "config.json"
    api_url = args.api_url or ""
    api_key = args.api_key or ""

    if needs_config and not api_url:
        if not config_path.exists():
            print("Error: ~/.promptbook/config.json not found. Run the Promptbook setup first.", file=sys.stderr)
            sys.exit(1)
        with open(config_path) as f:
            config = json.load(f)
        api_url = api_url or config.get("api_url", "")
        api_key = api_key or config.get("api_key", "")
        if not api_url or not api_key:
            print("Error: config.json missing api_url or api_key", file=sys.stderr)
            sys.exit(1)

    # Find and parse sessions
    if not args.json:
        print(f"Scanning for sessions from the last {args.days} days...", file=sys.stderr)

    jsonl_files = find_jsonl_files(args.days, args.project, args.before)

    if not args.json:
        print(f"Found {len(jsonl_files)} session files", file=sys.stderr)

    if not jsonl_files:
        if args.json:
            print("[]")
        else:
            print("Nothing to backfill.", file=sys.stderr)
        return

    # Parse all sessions
    sessions = []
    skipped = 0
    for jsonl_path in jsonl_files:
        session_data = parse_session(jsonl_path)
        if session_data is None:
            skipped += 1
        else:
            sessions.append(session_data)

    if not args.json:
        print(f"Parsed {len(sessions)} sessions ({skipped} skipped)", file=sys.stderr)

    # --json mode: output to stdout and exit
    if args.json:
        json.dump(sessions, sys.stdout, default=str)
        return

    # --dry-run mode: print summary
    if args.dry_run:
        for i, s in enumerate(sessions, 1):
            print(f"  [{i}/{len(sessions)}] {s['project_name']}: "
                  f"{s['prompt_count']} prompts, "
                  f"{s['total_tokens']:,} tokens, "
                  f"{s['build_time_seconds']}s, "
                  f"model={s['model']}")
        print(f"\nDone: {len(sessions)} sessions found, {skipped} empty/invalid")
        return

    # Default mode: upload to backfill batch endpoint
    if not sessions:
        print("No sessions to upload.", file=sys.stderr)
        return
    batch_id = upload_batch(sessions, api_url, api_key, args.days)
    if batch_id:
        print(f"Batch {batch_id}: {len(sessions)} sessions uploaded", file=sys.stderr)
        print(batch_id)  # stdout: just the batch_id for script consumption
    else:
        print("Upload failed — no batch created.", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: plugins/promptbook/scripts/backfill-history.js
#!/usr/bin/env node
/**
 * Promptbook — Historic Data Sync (Backfill)
 *
 * Scans local Claude Code JSONL transcripts and submits aggregate stats
 * as backfill Build Cards. Uses the same parsing logic as the live hooks.
 *
 * Never sends prompts, code, or file contents — only aggregate stats.
 *
 * Usage:
 *   node backfill-history.js [--days N] [--dry-run] [--json] [--project DIR]
 *                            [--before DATE] [--generate-summaries]
 *                            [--api-url URL] [--api-key KEY]
 */
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const { execSync, execFileSync } = require('child_process');
const { parseTranscript } = require('./lib/transcript');
const { getPrimaryLanguage, deriveProjectName } = require('./lib/language');
const { buildSummaryPrompt, generateFallbackTitle, generateFallbackSummary } = require('./lib/summary');

// --- CLI argument parsing ---
const args = process.argv.slice(2);
function getArg(name, defaultVal) {
  const idx = args.indexOf(`--${name}`);
  if (idx === -1) return defaultVal;
  if (typeof defaultVal === 'boolean') return true;
  return args[idx + 1] || defaultVal;
}
const DAYS = Number(getArg('days', '30'));
const DRY_RUN = getArg('dry-run', false);
const JSON_MODE = getArg('json', false);
const PROJECT_FILTER = getArg('project', '');
const BEFORE = getArg('before', '');
const GENERATE_SUMMARIES = getArg('generate-summaries', false);
const API_URL_ARG = getArg('api-url', '');
const API_KEY_ARG = getArg('api-key', '');

// --- Helpers ---
function log(msg) {
  process.stderr.write(msg + '\n');
}

/**
 * Count real user prompts in a JSONL file.
 * Only counts type=user entries with actual text content (not empty tool results).
 */
function countPrompts(jsonlPath) {
  let count = 0;
  const content = fs.readFileSync(jsonlPath, 'utf8');
  for (const line of content.split('\n')) {
    if (!line.trim()) continue;
    let entry;
    try { entry = JSON.parse(line); } catch { continue; }
    if (entry.type !== 'user') continue;
    const msg = entry.message || {};
    const c = typeof msg === 'object' ? msg.content : undefined;
    if (typeof c === 'string' && c.trim().length > 0) {
      count++;
    } else if (Array.isArray(c)) {
      if (c.some(b => b && b.type === 'text' && typeof b.text === 'string' && b.text.trim().length > 0)) {
        count++;
      }
    }
  }
  return count;
}

/**
 * Extract timestamps from JSONL entries for session duration.
 */
function extractTimestamps(jsonlPath) {
  const timestamps = [];
  const content = fs.readFileSync(jsonlPath, 'utf8');
  for (const line of content.split('\n')) {
    if (!line.trim()) continue;
    let entry;
    try { entry = JSON.parse(line); } catch { continue; }
    const ts = entry.timestamp;
    if (ts && typeof ts === 'string') {
      const d = new Date(ts);
      if (!isNaN(d.getTime())) timestamps.push(d);
    }
  }
  return timestamps;
}

/**
 * Count lines changed from tool_use blocks in the JSONL (same logic as track-edits.js).
 */
function countLinesChanged(jsonlPath) {
  let linesChanged = 0;
  const content = fs.readFileSync(jsonlPath, 'utf8');
  for (const line of content.split('\n')) {
    if (!line.trim()) continue;
    let entry;
    try { entry = JSON.parse(line); } catch { continue; }
    if (entry.type !== 'assistant') continue;
    const msg = entry.message || {};
    const blocks = msg.content;
    if (!Array.isArray(blocks)) continue;
    for (const block of blocks) {
      if (!block || block.type !== 'tool_use') continue;
      const toolName = block.name || '';
      const toolInput = block.input || {};
      if (toolName === 'Write') {
        const c = toolInput.content || '';
        if (c) linesChanged += c.split('\n').length;
      } else if (toolName === 'Edit') {
        const newStr = toolInput.new_string || '';
        const oldStr = toolInput.old_string || '';
        if (newStr) {
          const diff = Math.abs(newStr.split('\n').length - (oldStr ? oldStr.split('\n').length : 0));
          linesChanged += diff || (newStr ? 1 : 0);
        }
      } else if (toolName === 'NotebookEdit') {
        const src = toolInput.new_source || '';
        linesChanged += src ? src.split('\n').length : 1;
      }
    }
  }
  return linesChanged;
}

/**
 * Find JSONL session files from the last N days.
 */
function findJsonlFiles() {
  const claudeDir = path.join(os.homedir(), '.claude', 'projects');
  if (!fs.existsSync(claudeDir)) {
    log(`No Claude Code projects found at ${claudeDir}`);
    return [];
  }

  const afterCutoff = Date.now() - (DAYS * 86400 * 1000);
  const beforeCutoff = BEFORE ? new Date(BEFORE).getTime() : null;
  const files = [];

  for (const dirName of fs.readdirSync(claudeDir)) {
    const projectDir = path.join(claudeDir, dirName);
    if (!fs.statSync(projectDir).isDirectory()) continue;

    if (PROJECT_FILTER) {
      const decoded = dirName.split('-').filter(Boolean).pop() || dirName;
      if (!dirName.toLowerCase().includes(PROJECT_FILTER.toLowerCase()) &&
          decoded.toLowerCase() !== PROJECT_FILTER.toLowerCase()) {
        continue;
      }
    }

    for (const file of fs.readdirSync(projectDir)) {
      if (!file.endsWith('.jsonl')) continue;
      const filePath = path.join(projectDir, file);
      try {
        const stat = fs.statSync(filePath);
        if (!stat.isFile()) continue;
        const mtime = stat.mtimeMs;
        if (mtime >= afterCutoff && (!beforeCutoff || mtime < beforeCutoff)) {
          files.push(filePath);
        }
      } catch { continue; }
    }
  }

  return files.sort((a, b) => fs.statSync(a).mtimeMs - fs.statSync(b).mtimeMs);
}

/**
 * Parse a single JSONL session file into a build record.
 * Uses the shared lib/ modules (same code as live hooks).
 */
function parseSession(jsonlPath) {
  const sessionId = path.basename(jsonlPath, '.jsonl');
  const projectDirName = path.basename(path.dirname(jsonlPath));

  // Decode the Claude Code directory name back to a real path, then use shared
  // deriveProjectName() which handles git worktrees.
  // Format: -Users-foo-Desktop-myproject → /Users/foo/Desktop/myproject
  const decodedPath = '/' + projectDirName.replace(/^-/, '').replace(/-/g, '/');
  const projectName = deriveProjectName(decodedPath);

  // Extract timestamps for session duration
  const timestamps = extractTimestamps(jsonlPath);
  if (timestamps.length < 2) return null;

  const startTime = new Date(Math.min(...timestamps.map(t => t.getTime())));
  const endTime = new Date(Math.max(...timestamps.map(t => t.getTime())));
  const buildTimeSeconds = Math.round((endTime - startTime) / 1000);

  if (buildTimeSeconds < 5) return null;

  // Count real user prompts (same logic as UserPromptSubmit hook)
  const promptCount = countPrompts(jsonlPath);

  // Parse transcript using shared lib (same code as session-end.js)
  const transcript = parseTranscript(jsonlPath, '');
  if (transcript.error) return null;
  if (!transcript.model) return null;

  const totalTokens = transcript.total_tokens || 0;

  // Skip noise (same filter as session-end.js)
  if (promptCount === 0 && buildTimeSeconds < 10 && totalTokens < 1000) return null;
  if (totalTokens === 0 && promptCount === 0) return null;

  // Count lines changed (same logic as track-edits.js)
  const linesChanged = countLinesChanged(jsonlPath);

  // Language detection using shared lib (same as session-end.js)
  const fileExtensions = transcript.source_metadata?.file_extensions || {};
  const language = getPrimaryLanguage(fileExtensions);

  // Tool summary for fallback title generation
  const toolSummary = transcript.tool_usage_summary || {};
  const toolSummaryStr = Object.entries(toolSummary)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 3)
    .map(([k, v]) => `${v} ${k}`)
    .join(', ');

  return {
    session_id: sessionId,
    project_name: projectName,
    model: transcript.model,
    ai_tool: 'claude-code',
    start_time: startTime.toISOString(),
    end_time: endTime.toISOString(),
    build_time_seconds: buildTimeSeconds,
    prompt_count: promptCount,
    lines_changed: linesChanged || null,
    language,
    status: 'completed',
    total_tokens: totalTokens,
    input_tokens: transcript.input_tokens || 0,
    output_tokens: transcript.output_tokens || 0,
    cache_creation_input_tokens: transcript.cache_creation_input_tokens || 0,
    cache_read_input_tokens: transcript.cache_read_input_tokens || 0,
    source: 'backfill',
    source_metadata: {
      file_extensions: fileExtensions,
      tool_usage_summary: toolSummary,
    },
    // Kept locally for summary generation — stripped before upload
    _compact_log: transcript.compact_log || '',
    _tool_summary_str: toolSummaryStr,
    _jsonl_path: jsonlPath,
  };
}

/**
 * Generate a title and summary for a session using Haiku (or fallback).
 * Same logic as submit.js generateAndPublish().
 */
function generateSummary(session) {
  let compactLog = session._compact_log || '';
  if (!compactLog) return null;

  // Truncate to ~4000 chars
  if (compactLog.length > 4000) {
    compactLog = compactLog.slice(0, 4000) + '\n... (truncated)';
  }

  // Try Haiku via claude CLI
  let claudeCmd = '';
  try {
    claudeCmd = execSync('which claude 2>/dev/null || where claude 2>nul', { encoding: 'utf8' }).trim();
  } catch { /* not found */ }

  if (claudeCmd) {
    try {
      const prompt = buildSummaryPrompt(compactLog, session.project_name, session.prompt_count);
      const response = execFileSync(claudeCmd, ['--print', '--model', 'haiku'], {
        input: prompt,
        encoding: 'utf8',
        timeout: 60000,
        env: { ...process.env, PROMPTBOOK_SKIP_HOOKS: '1' },
      });

      if (response) {
        const titleMatch = response.match(/^TITLE:\s*(.+)/im);
        const summaryMatch = response.match(/^SUMMARY:\s*(.+)/im);
        if (titleMatch || summaryMatch) {
          return {
            title: titleMatch ? titleMatch[1].trim() : '',
            summary: summaryMatch ? summaryMatch[1].trim() : '',
          };
        }
      }
    } catch (err) {
      log(`  Haiku failed for ${session.session_id.slice(0, 8)}...: ${err.message}`);
    }
  }

  // Fallback: deterministic title/summary (same as submit.js)
  return {
    title: generateFallbackTitle(
      session.prompt_count, session.build_time_seconds,
      session.project_name, session._tool_summary_str
    ),
    summary: generateFallbackSummary(
      session.prompt_count, session.build_time_seconds,
      session.total_tokens, session.model, session._tool_summary_str
    ),
  };
}

/**
 * Upload sessions to the backfill batch endpoint.
 */
async function uploadBatch(sessions, apiUrl, apiKey) {
  const chunkSize = 50;
  let batchId = null;

  for (let i = 0; i < sessions.length; i += chunkSize) {
    const chunk = sessions.slice(i, i + chunkSize);

    // Strip internal fields before upload
    const cleaned = chunk.map(s => {
      const { _compact_log, _tool_summary_str, _jsonl_path, ...rest } = s;
      return rest;
    });

    try {
      const response = await fetch(`${apiUrl}/api/backfill/upload`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${apiKey}`,
        },
        body: JSON.stringify({ days_scanned: DAYS, sessions: cleaned }),
        signal: AbortSignal.timeout(30000),
      });

      if (!response.ok) {
        const body = await response.text();
        log(`  Upload failed (HTTP ${response.status}): ${body}`);
        if ([401, 403, 500].includes(response.status)) {
          log('  Stopping due to server error.');
          break;
        }
        continue;
      }

      const body = await response.json();
      if (!batchId) batchId = body.batch_id;
      log(`  Uploaded ${body.uploaded || 0} sessions (${body.duplicates || 0} duplicates)`);
    } catch (err) {
      log(`  Upload failed: ${err.message}`);
      break;
    }

    // Brief pause between chunks for rate limiting
    if (i + chunkSize < sessions.length) {
      await new Promise(r => setTimeout(r, 2000));
    }
  }

  return batchId;
}

// --- Main ---
async function main() {
  // Load config
  let apiUrl = API_URL_ARG;
  let apiKey = API_KEY_ARG;
  const needsConfig = !DRY_RUN && !JSON_MODE;

  if (needsConfig && !apiUrl) {
    const configPath = path.join(os.homedir(), '.promptbook', 'config.json');
    if (!fs.existsSync(configPath)) {
      log('Error: ~/.promptbook/config.json not found. Run the Promptbook setup first.');
      process.exit(1);
    }
    const config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
    apiUrl = apiUrl || config.api_url || '';
    apiKey = apiKey || config.api_key || '';
    if (!apiUrl || !apiKey) {
      log('Error: config.json missing api_url or api_key');
      process.exit(1);
    }
  }

  // Find sessions
  if (!JSON_MODE) log(`Scanning for sessions from the last ${DAYS} days...`);
  const jsonlFiles = findJsonlFiles();
  if (!JSON_MODE) log(`Found ${jsonlFiles.length} session files`);

  if (jsonlFiles.length === 0) {
    if (JSON_MODE) process.stdout.write('[]');
    else log('Nothing to backfill.');
    return;
  }

  // Parse all sessions
  const sessions = [];
  let skipped = 0;
  for (const filePath of jsonlFiles) {
    const session = parseSession(filePath);
    if (session) sessions.push(session);
    else skipped++;
  }

  if (!JSON_MODE) log(`Parsed ${sessions.length} sessions (${skipped} skipped)`);

  // Generate summaries if requested
  if (GENERATE_SUMMARIES && sessions.length > 0) {
    log(`Generating summaries for ${sessions.length} sessions...`);
    for (let i = 0; i < sessions.length; i++) {
      const s = sessions[i];
      const result = generateSummary(s);
      if (result) {
        s.title = result.title;
        s.summary = result.summary;
        if (!JSON_MODE) {
          const label = result.title.length > 50 ? result.title.slice(0, 50) + '...' : result.title;
          log(`  [${i + 1}/${sessions.length}] ${label}`);
        }
      }
    }
  }

  // --json mode
  if (JSON_MODE) {
    const cleaned = sessions.map(s => {
      const { _compact_log, _tool_summary_str, _jsonl_path, ...rest } = s;
      return rest;
    });
    process.stdout.write(JSON.stringify(cleaned));
    return;
  }

  // --dry-run mode
  if (DRY_RUN) {
    for (let i = 0; i < sessions.length; i++) {
      const s = sessions[i];
      log(`  [${i + 1}/${sessions.length}] ${s.project_name}: ` +
        `${s.prompt_count} prompts, ${s.total_tokens.toLocaleString()} tokens, ` +
        `${s.build_time_seconds}s, ${s.lines_changed || 0} lines, model=${s.model}`);
    }
    log(`\nDone: ${sessions.length} sessions found, ${skipped} empty/invalid`);
    return;
  }

  // Upload
  if (sessions.length === 0) {
    log('No sessions to upload.');
    return;
  }

  const batchId = await uploadBatch(sessions, apiUrl, apiKey);
  if (batchId) {
    log(`Batch ${batchId}: ${sessions.length} sessions uploaded`);
    process.stdout.write(batchId); // stdout: just the batch_id for script consumption
  } else {
    log('Upload failed — no batch created.');
    process.exit(1);
  }
}

main().catch(err => {
  log(`Fatal: ${err.message}`);
  process.exit(1);
});
